### PR TITLE
Add section on ImmutableJS

### PR DESCRIPTION
**Pull request by jackfranklin**
Thoughts @english @harrison @maxmurdoch please

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -660,6 +660,74 @@ angular.module('AlertListComponentModule', [])
...
+  function present(events) {
+    return Immutable.fromJS(events).map(function(event) {
+      event.set('some_data', true);
+    }).map(function(event) {
+      ...
+    }).toJS();
+  }
+
+  return {
+    events: events
```
Comment (harrison) [inline README.md]: `present`

File: README.md [hunk 2]
```diff
@@ -660,6 +660,74 @@ angular.module('AlertListComponentModule', [])
...
 
+### Services and Factories
+
+#### Services should always export an object, not just a single function
+
+_Why_: easier to see at the definition site and the call site exactly what function the service provides
+
+```js
+// Recommended
+angular.module('BuildCSVService', []).factory('BuildCSV', function() {
```
Comment (harrison) [inline README.md]: `buildCSVModule`

File: README.md [hunk 3]
```diff
@@ -660,6 +660,74 @@ angular.module('AlertListComponentModule', [])
...
+    // ...
+  }
+
+  return build;
+});
+```
+
+#### Services that take in data and manipulate it should never mutate the original object and return the new object
+
+__Why__: Makes code easier to reason about when it's clearer what is and isn't being mutated.
```
Comment (harrison) [inline README.md]: Maybe something like `Avoiding mutation in service objects makes it possible to reason about them from their call site without knowing what they do internally.`

File: README.md [hunk 4]
```diff
@@ -660,6 +660,74 @@ angular.module('AlertListComponentModule', [])
...
 
+### Services and Factories
+
+#### Services should always export an object, not just a single function
+
+_Why_: easier to see at the definition site and the call site exactly what function the service provides
+
+```js
+// Recommended
+angular.module('buildCSVModule', []).factory('BuildCSV', function() {
```
Comment (maxmurdoch) [inline README.md]: name func

File: README.md [hunk 5]
```diff
@@ -660,6 +660,74 @@ angular.module('AlertListComponentModule', [])
...
+  function build() {
+    // ...
+  }
+
+  return {
+    build: build
```
Comment (maxmurdoch) [inline README.md]: trailing

**Comment by english** [comment]
Looks good to me :+1:

File: README.md [hunk 6]
```diff
@@ -660,6 +660,74 @@ angular.module('AlertListComponentModule', [])
...
+    }).toJS();
+  }
+
+  return {
+    present: present
+  }
+});
+```
+
+
```
Comment (harrison) [inline README.md]: :scissors:

File: README.md [hunk 7]
```diff
@@ -660,6 +660,74 @@ angular.module('AlertListComponentModule', [])
 - Don’t manipulate DOM in your controllers, this will make them harder to test. Use directives instead.
 
 
+### Services and Factories
+
+#### Services should always export an object, not just a single function
```
Comment (harrison) [inline README.md]: Could maybe be re-worded, not sure exactly to what. `Treat Service objects like a Class with static methods; don't export services as a single function`

File: README.md [hunk 8]
```diff
@@ -660,6 +660,74 @@ angular.module('AlertListComponentModule', [])
...
+
+_Why_: easier to see at the definition site and the call site exactly what function the service provides
```
Comment (maxmurdoch) [inline README.md]: full stop

File: README.md [hunk 9]
```diff
@@ -660,6 +660,74 @@ angular.module('AlertListComponentModule', [])
...
+
+```js
+// Recommended
```
Comment (maxmurdoch) [inline README.md]: use 'Recommend/Avoid' in all examples

File: README.md [hunk 10]
```diff
@@ -660,6 +660,74 @@ angular.module('AlertListComponentModule', [])
...
+
+// Avoid
+var events = [...];
+EventPresenterService.present(events);
+// events has been mutated
+```
+
+#### Prefer [ImmutableJS](https://github.com/facebook/immutable-js) when creating services that manipulate data
+
+Use `fromJS` to convert a JS object into an Immutable type, and then `toJS()` at the end to convert back.
```
Comment (maxmurdoch) [inline README.md]: remove superfluous parens after `toJS`

File: README.md [hunk 11]
```diff
@@ -660,6 +660,74 @@ angular.module('AlertListComponentModule', [])
...
+
+__Why__: ImmutableJS is a fantastic library for taking data and manipulating it without ever mutating.
+
+```js
+// Recommend
+angular.module('eventPresenterModule', []).factory('EventPresenterService', function() {
```
Comment (maxmurdoch) [inline README.md]: name factory func

File: README.md [hunk 12]
```diff
@@ -660,6 +660,74 @@ angular.module('AlertListComponentModule', [])
...
+  function present(events) {
+    return Immutable.fromJS(events).map(function(event) {
+      event.set('some_data', true);
+    }).map(function(event) {
```
Comment (maxmurdoch) [inline README.md]: name functions

File: README.md [hunk 13]
```diff
@@ -660,6 +660,74 @@ angular.module('AlertListComponentModule', [])
...
+      ...
+    }).toJS();
+  }
+
+  return {
+    present: present
```
Comment (maxmurdoch) [inline README.md]: trailing

**Pull request merged**